eftest/src/eftest/runner.clj

```diff
@@ -40,14 +41,16 @@
       result)))
 
 (defn- run-in-parallel [fs]
-  (let [executor (java.util.concurrent.Executors/newFixedThreadPool
-                   (+ 2 (.availableProcessors (Runtime/getRuntime))))]
+  (let [^ExecutorService executor (java.util.concurrent.Executors/newFixedThreadPool
+                                  (+ 2 (.availableProcessors (Runtime/getRuntime))))]
     (try
-      (->> fs
-           (map #(.submit executor ^Runnable %))
-           doall
+      (->>
+          (for [f fs
+                :let [^Callable cb (cast Callable (bound-fn [] (f)))]]
+                (.submit executor cb))
+           (doall)
            (map #(.get %))
-           dorun)
+           (doall))
```

Comment:
I think if we're going to change this from `dorun` to `doall`, the function name shouldn't be called `run-in-parallel`. Instead, maybe we want something more like:

```clojure
(defn- ^Callable bound-callback [f]
  (cast Callable (bound-fn* f)))

(defn- ^ExecutorService threadpool-executor []
  (Executors/newFixedThreadPool (+ 2 (.availableProcessors (Runtime/getRuntime))))

(defn- pcalls* [fs]
  (let [executor (threadpool-executor)]
    (->> fs
         (map #(.submit executor (bound-callback %)))
         (doall)
         (map #(.get %)))))

(defn- pmap* [f xs]
  (pcalls* (map (fn [x] #(f x)) xs)))

(defn- test-all [vars opts]
  (let [mapf (if (:multithread-ns? opts true) pmap* map)]
    (->> (group-by (comp :ns meta) vars)
         (mapf (fn [[ns vars]] (test-ns ns vars opts)))
         (apply merge-with +)))
```

Then we replace `run-in-parallel` with `pmap*` elsewhere as well.

Reply:
agree